### ISSUE TITLE
Fails to detect 2017 command line build tools

### ISSUE THREAD

**Issue by d3x0r** [cmake-js/cmake-js#100]
The registry key format and keys in the installed keys is different for visual studio 2017 (hooray for standardization huh?)  

Small modification to vsDetect.js fixes it.

working on a PR to attach to this.

**Comment by unbornchikken** [comment]
Since I'm not using that pile of crap (aka Windows) anymore for development, I need a little help on this. Do you see a good spot in the registry or anywhere to be able to detect build tools installed by this way, please?

**Comment by wfjsw** [comment]
Have you taken any look on https://github.com/Microsoft/vswhere/ ?

https://github.com/Microsoft/vswhere/wiki/Find-MSBuild

**Comment by d3x0r** [comment]
> which install MSBuild 15.8.169+g1ccb72aefa.
> 
> The toolscore registry is not present and both cmake-js and cmake failed to detect it.

Noticed MSBuild installed 2015, not 2017
There is another error that this may be for you...  node-gyp has similar issues sometimes after installing just command line tools on a clean windows system.

`npm config set msvs_version 2015 --global`

**Comment by unbornchikken** [comment]
> > which install MSBuild 15.8.169+g1ccb72aefa.
> > The toolscore registry is not present and both cmake-js and cmake failed to detect it.
> 
> Noticed MSBuild installed 2015, not 2017
> There is another error that this may be for you... node-gyp has similar issues sometimes after installing just command line tools on a clean windows system.
> 
> `npm config set msvs_version 2015 --global`

Yeah, it would be a good idea to use that if present.

**Comment by diegoperini** [comment]
> After I reinstalled the build tools from the official installer on ms website, CMake has successfully detected it.
> 
> Gotta try that in a fresh environment but I need time and hardware to prepare. I will keep this updated when I am ready.

This didn't work for me. Installing 2015 tools sounds intimidating, after setting up a working 2017 environment. Is there easy way to make it work with 2017 build tools?

**Comment by diegoperini** [comment]
I submitted https://github.com/cmake-js/cmake-js/pull/148 to fix this issue. The patch introduces `vswhere`. The problem is fixed on my fresh Windows 10, VS2017 machine.

**Comment by unbornchikken** [comment]
I'm wondering about how much is this issue important to get fixed for you guys? Because I have merged a semver major fix https://github.com/cmake-js/cmake-js/pull/151, and don't want to put on a new major release until my planned features get completed. But if it is important, I'm gonna release the fix for this only asap.

**Comment by diegoperini** [comment]
I would assume VS2017 is quite common among many developers. We didn't break any existing setup or didn't introduce a regression (probably). Can't we go with a minor release?

**Comment by wfjsw** [comment]
> > which install MSBuild 15.8.169+g1ccb72aefa.
> > The toolscore registry is not present and both cmake-js and cmake failed to detect it.
> 
> Noticed MSBuild installed 2015, not 2017
> There is another error that this may be for you... node-gyp has similar issues sometimes after installing just command line tools on a clean windows system.
> 
> `npm config set msvs_version 2015 --global`

Nah the Visual Studio 2017 is actually v15 🤷‍♂️

**Comment by d3x0r** [comment]
I can wait 
but for a feature request (and disrelated to this)  I would like the ability to specify build 'RelWithDebInfo'  instead of just 'Release' and 'Debug'  I can specify --CDCMAKE_BUILD_TYPE=RelWithDebInfo, but the build still runs `cmake --build --config "Release"` or Debug

**Comment by diegoperini** [comment]
[This](https://github.com/MichaReiser/llvm-node) repo currently depends on cmake-js and fails to build on Windows due this issue, which is kind of important for me and maybe for a few other people as well.

**Comment by unbornchikken** [comment]
It's just not a CMake way, it's a node-gyp way. I would be much better is one can specify **CMake generator** to use in npm config. Since we have an option for that, the easiest possible solution would be a way to specify default CMake.js options with npm config:

```bash
npm config set cmake_js="-G='Visual Studio 1544 Win64'"
```

And for the automatic way, I think we can use this thing instead of registry hacking:

https://github.com/Microsoft/vswhere/releases

The only problem atm is in GRAPHISOFT we've been making or new cloud platform release, and I simply don't have enough strength in the evenings to make this happen. So I can offer a bright spot in the credit list for anyone who can help me on this. Or please give me a couple of weeks to get back to this project.

**Comment by unbornchikken** [comment]
Ok, I had some time and put together a draft of the new release. I think https://github.com/cmake-js/cmake-js/pull/151 is not worth a new semver major, even if it could lead a breaking change in some circumstances. I decided to not use https://github.com/cmake-js/cmake-js/pull/148, because adding executables to an npm module is an evil thing. I'm thinking on an auto download solution.

So the next feature that might solve this issue is the following proposal:

You can set up default options for CMake.js by using npm config by using the following pattern:

```bash
npm config set cmake_js_G "Visual Studio 56 Win128"
```

The option is the part of a **cmake_js_** pattern, that can be anything that cmake-js --help gives, eg:

cmake_js_G: you can set the generator
cmake_js_log-level: log level to use

So the solution for this issue would be to set the appropriate generator to use with npm config. You can see the list of known generators by entering:

```bash
cmake --help
```

Please test master branch and provide feedback!

**Comment by diegoperini** [comment]
Quote from https://github.com/Microsoft/vswhere/wiki/Installing
> Starting with Visual Studio 15.2 (26418.1 Preview) vswhere.exe is installed in %ProgramFiles(x86)%\Microsoft Visual Studio\Installer (use %ProgramFiles% in a 32-bit program prior to Windows 10). This is a fixed location that will be maintained.

Maybe this can help with the exe issue.

**Comment by unbornchikken** [comment]
Good catch. Definitely. It's good to see that they are seeing our pain. Windows tooling support is a freaking mess. They have been working hard to prevent developers to make programs to their platform.

**Comment by unbornchikken** [comment]
The prototype version has been released in 5.0.0. I'm waiting for a PR regarding to use "%ProgramFiles(x86)%\Microsoft Visual Studio\Installer" path, since I cannot test it properly on my system. Until then you can use:

```
npm config set cmake_js_G "<ganarator>"
```
